### PR TITLE
GH-40693: [Go] Fix Decimal type precision loss on GetOneForMarshal

### PR DESCRIPTION
**Pull request by disq**
### Rationale for this change

Loss of precision when using `GetOneForMarshal` on `Decimal128` and `Decimal256`

### What changes are included in this PR?

Fixes for precision loss with `DecimalType.GetOneForMarshal`

* GitHub Issue: #40693

### CODE REVIEW

**Comment by disq** [comment]
This seems to work:
```go
	typ := a.DataType().(*arrow.Decimal256Type)
	f := (&big.Float{}).SetInt(a.Value(i).BigInt())
	scale := big.NewInt(10)
	scale.Exp(scale, big.NewInt(int64(typ.Scale)), nil)
	f.Quo(f, new(big.Float).SetInt(scale))
	return f.Text('f', int(typ.Scale))
	// return f.Text('g', int(typ.Precision))
```

But the `f` format is not what the tests expect... with `g` it results in further floaty errors, as in `0.99000000000000000000000000827180612553027674871409` or `1234567890.1234567889999999999999999938549877259069`

**Comment by disq** [comment]
@hermanschaaf Came through and fixed these 🥇

**Comment by disq** [comment]
@zeroshade PTAL when you have a chance? Thanks!

File: go/arrow/array/decimal128.go [hunk 1]
```diff
@@ -86,15 +85,19 @@ func (a *Decimal128) setData(data *Data) {
...
-	f := (&big.Float{}).SetInt(a.Value(i).BigInt())
-	f.Quo(f, big.NewFloat(math.Pow10(int(typ.Scale))))
+	n := a.Value(i)
+	scale := typ.Scale
+	f := (&big.Float{}).SetInt(n.BigInt())
+	if scale < 0 {
+		f.SetPrec(128).Mul(f, (&big.Float{}).SetInt(decimal128.GetScaleMultiplier(int(-scale)).BigInt()))
+	} else {
+		f.SetPrec(128).Quo(f, (&big.Float{}).SetInt(decimal128.GetScaleMultiplier(int(scale)).BigInt()))
+	}
```
Comment (zeroshade) [inline go/arrow/array/decimal128.go:100]: huh, didn't expect the solution to be so "simple". Very nice

**Review by zeroshade — APPROVED** [review]
This looks great, thanks so much. The solution was easier than I was expecting

**Pull request merged**